### PR TITLE
Implement 2D ellipse footprint decomposition

### PR DESCRIPTION
**Pull request by grlee77**
## Description

related to #5966 and #5650

This PR builds upon #5482 and #6150 (initial commits here are the same), implementing a decomposition of 2D disk and elliptical structuring elements using (possibly elongated) cross-shaped structuring elements. The algorithm was implemented by adapting pseudo-code from the following reference

```
Li, D. and Ritter, G.X.
Decomposition of Separable and Symmetric Convex Templates.
Proc. SPIE 1350, Image Algebra and Morphological Image Processing, (1 November 1990).
https://doi.org/10.1117/12.23608
```

This provides an alternative to the disk decomposition proposed in #6150 and is applicable to arbitrary radius as it does not rely on precomputing optimal combinations of structuring elements. In benchmarks it is slightly slower than the "optimal" one in #6150. The primary benefit of the decomposition in this PR is that it can use elongated elements, allowing approximation to ellipses in addition to disks. I could not come up with a simple adaptation to 3D, though, so there is no complement here to the `ball` decomposition into 3x3x3 elements given in #6150.

### Benchmark results
```
[ 75.00%] ··· benchmark_morphology.BinaryMorphology2D.time_erosion                                                                                                          ok
[ 75.00%] ··· ============ =========== ======== ============= ============= ============= =============
              --                                                     decomposition                    
              --------------------------------- -------------------------------------------------------
                 shape      footprint   radius       None        sequence     separable      crosses  
              ============ =========== ======== ============= ============= ============= =============
               (512, 512)      disk       1      2.84±0.02ms   3.68±0.02ms       n/a       2.84±0.01ms
               (512, 512)      disk       3       7.66±0.2ms   6.55±0.03ms       n/a       6.62±0.02ms
               (512, 512)      disk       5      17.7±0.05ms   6.62±0.04ms       n/a       10.2±0.06ms
               (512, 512)      disk       15      141±0.7ms    18.3±0.07ms       n/a       18.2±0.04ms
               (512, 512)      disk       25       380±5ms      18.1±0.1ms       n/a       19.6±0.09ms
               (512, 512)      disk       40       966±8ms     18.5±0.05ms       n/a       32.7±0.06ms
               (512, 512)    ellipse      1      3.68±0.02ms       n/a           n/a       4.98±0.02ms
               (512, 512)    ellipse      3      9.94±0.03ms       n/a           n/a       8.57±0.01ms
               (512, 512)    ellipse      5      24.1±0.04ms       n/a           n/a       9.37±0.03ms
               (512, 512)    ellipse      15       157±2ms         n/a           n/a       17.8±0.04ms
               (512, 512)    ellipse      25       399±5ms         n/a           n/a       25.2±0.03ms
               (512, 512)    ellipse      40       992±10ms        n/a           n/a        31.9±0.2ms
```

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+"""
+=======================================================
+Decomposition of flat footprints (structuring elements)
+=======================================================
+
+Many footprints (structuring elements) can be decomposed into an equivalent
+series of smaller structuring elements. The term "flat" refers to a footprints
+that only contains values of 0 or 1 (i.e. all methods in
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:8]:
```suggestion
series of smaller structuring elements. The term "flat" refers to footprints
that only contain values of 0 or 1 (i.e., all methods in
```

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+``skimage.morphology.footprints``). Binary dilation operations have an
+associative and distributive property such that often allows decomposition into
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:10]:
```suggestion
associative and distributive property that often allows decomposition into
```

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+an equivalent series of smaller footprints. Most often this is done to provide
+a performance benefit.
+
+As a concrete example, dilation with a square footprint of size (15, 15) is
+equivalent to dilation with a rectangle of size (15, 1) followed by a
+dilation with a rectangle of size (1, 15). It is also equivalent to 7
+consecutive dilations with a shape (3, 3) square.
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:17]:
```suggestion
consecutive dilations with a square footprint of size (3, 3).
```
?

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+Many footprints (structuring elements) can be decomposed into an equivalent
+series of smaller structuring elements. The term "flat" refers to a footprints
+that only contains values of 0 or 1 (i.e. all methods in
+``skimage.morphology.footprints``). Binary dilation operations have an
+associative and distributive property such that often allows decomposition into
+an equivalent series of smaller footprints. Most often this is done to provide
+a performance benefit.
+
+As a concrete example, dilation with a square footprint of size (15, 15) is
+equivalent to dilation with a rectangle of size (15, 1) followed by a
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:15]:
```suggestion
equivalent to dilation with a rectangle of size (15, 1) followed by another
```

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+dilation with a rectangle of size (1, 15). It is also equivalent to 7
+consecutive dilations with a shape (3, 3) square.
+
+There are many possible decompositions and which one performs best may be
+architecture-dependent.
+
+scikit-image currently provides two forms of automated decomposition. For the
+cases of ``square``, ``rectangle`` and ``cube`` footprints, there is an option
+for a "separable" decomposition (size > 1 along only 1 axis at a time).
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:24]:
```suggestion
for a "separable" decomposition (size > 1 along only one axis at a time).
```

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+For some other symmetric convex shapes such as ``diamond``, ``octahedron`` and
+``octagon`` there is no separable decomposition, but it is possible to provide
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:27]:
```suggestion
``octagon``, there is no separable decomposition, but it is possible to provide
```

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+a "sequence" decomposition based on a series of small footprints of shape
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:28]: Use 'shape' also in the second paragraph (see above)?

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+``(3,) * ndim``.
+
+For simplicity of implementation, all decompositions use only odd-sized
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:31]:
```suggestion
For simplicity of implementation, all decompositions shown below use only odd-sized
```
?

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
+"""
+=======================================================
+Decomposition of flat footprints (structuring elements)
```
Comment (mkcor) [inline doc/examples/numpy_operations/plot_structuring_element_decompositions.py:3]:
```suggestion
Decompose flat footprints (structuring elements)
```

**Comment by alexdesiqueira** [comment]
@grlee77 could you give this some love according to @mkcor's comments? :grin: Thanks for your work here, and in #5482 #6150 also, Greg! I think we can merge this series of puppies very soon :slightly_smiling_face:

File: doc/examples/numpy_operations/plot_structuring_element_decompositions.py (hunk 6 above)
Comment (alexdesiqueira) [reply doc/examples/numpy_operations/plot_structuring_element_decompositions.py:27]: How about:
```
There is no separable decomposition for some other symmetric convex shapes,
e.g., ``diamond``, ``octahedron`` and ``octagon``. However, it is possible to provide
```

**Comment by grlee77** [comment]
> @grlee77 could you give this some love according to @mkcor's comments? grin Thanks for your work here, and in https://github.com/scikit-image/scikit-image/pull/5482 https://github.com/scikit-image/scikit-image/pull/6150 also, Greg! I think we can merge this series of puppies very soon slightly_smiling_face

I have made the suggested changes. Let me fix the conflicts now...

**Comment by grlee77** [comment]
also cherry-picked a few commits corresponding to reviewer feedback in #6150 here. At this point, it might be easier to just review everything together (`disk`, `ball` and `ellipse`) in this PR and we can close #6150?

**Comment by alexdesiqueira** [comment]
Thanks @grlee77!
> I think we can merge this series of puppies very soon :slightly_smiling_face:

I wonder why I referred to these as puppies... :joy: Maybe because of the last meeting we had!

**Review by alexdesiqueira — APPROVED** [review]

**Comment by alexdesiqueira** [comment]
> At this point, it might be easier to just review everything together (disk, ball and ellipse) in this PR and we can close https://github.com/scikit-image/scikit-image/pull/6150?

@grlee77 your call. I think this one is good to go, though! I can help you with #6150 later, if you'd like.

**Comment by alexdesiqueira** [comment]
Great stuff @grlee77 @rfezzani @mkcor. Thank you!

**Pull request merged**